morpheus/_lib/src/stages/inference_client_stage.cpp

```diff
@@ -0,0 +1,276 @@
+{
+    int32_t retry_count = 0;
+
+    using namespace std::chrono_literals;
+
+    auto backoff = ExponentialBackoff(on, 100ms, 4000ms);
+
+    while (true)
+    {
+        try
+        {
+            // Using the `count` which is the number of rows in the inference tensors. We will check later if this
+            // doesn't match the number of rows in the dataframe (`mess_count`). This happens when the size of the
+            // input is too large and needs to be broken up in chunks in the pre-process stage. When this is the
+            // case we will reduce the rows in the response outputs such that we have a single response for each
+            // row int he dataframe.
+            // TensorMap output_tensors;
+            // buffer_map_t output_buffers;
+
+            auto lock = std::shared_lock(m_session_mutex);
+
+            auto session = m_client->get_session();
+
+            TensorMap model_input_tensors;
+
+            for (auto mapping : session->get_input_mappings(m_input_mapping))
+            {
+                if (x->memory->has_tensor(mapping.tensor_field_name))
+                {
+                    model_input_tensors[mapping.model_field_name].swap(x->get_input(mapping.tensor_field_name));
+                }
+            }
+
+            // TODO(cwharris): Break inference in to batches and attempt retries on per-batch basis.
+            auto model_output_tensors = co_await session->infer(std::move(model_input_tensors));
+
+            co_await on->yield();
+
+            if (x->mess_count != x->count)
+            {
+                reduce_outputs(x, model_output_tensors);
+            }
+
+            // If we need to do logits, do that here
+            if (m_needs_logits)
+            {
+                apply_logits(model_output_tensors);
+            }
+
+            TensorMap output_tensor_map;
+
+            for (auto mapping : session->get_output_mappings(m_output_mapping))
+            {
+                auto pos = model_output_tensors.find(mapping.model_field_name);
+
+                if (pos != model_output_tensors.end())
+                {
+                    output_tensor_map[mapping.tensor_field_name].swap(
+                        std::move(model_output_tensors[mapping.model_field_name]));
+
+                    model_output_tensors.erase(pos);
+                }
+            }
+
+            // Final output of all mini-batches
+            auto response_mem = std::make_shared<ResponseMemory>(x->mess_count, std::move(output_tensor_map));
+
+            auto response = std::make_shared<MultiResponseMessage>(
+                x->meta, x->mess_offset, x->mess_count, std::move(response_mem), 0, response_mem->count);
+
+            co_yield std::move(response);
+
+            co_return;
+
+        } catch (...)
+        {
+            auto lock = std::unique_lock(m_session_mutex);
+
+            this->m_client->reset_session();
```

Comment:
Something needs to keep track of what the current session is, and make that information shareable across multiple invocations `on_data`. Having the client keep track of that information is what's happening now. We could instead store the session on the stage, which might make more sense because we're also handling the locking here.